File: app/src/main/java/com/haishinkit/iso/AVCConfigurationRecord.java
package com.haishinkit.iso;

import android.media.MediaFormat;
import org.apache.commons.lang3.builder.ToStringBuilder;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

public final class AVCConfigurationRecord {
    public static final byte RESERVE_LENGTH_SIZE_MINUS_ONE = (byte) 0x3F;
    public static final byte RESERVE_NUM_OF_SEQUENCE_PARAMETER_SETS = (byte) 0xE0;
    public static final byte RESERVE_CHROME_FORMAT = (byte) 0xFC;
    public static final byte RESERVE_BIT_DEPTH_LUMA_MINUS8 = (byte) 0xF8;
    public static final byte RESERVE_BIT_DEPTH_CHROME_MINUS8 = (byte) 0xF8;

    private byte configurationVersion = 1;
    private byte AVCProfileIndication = 0;
    private byte profileCompatibility = 0;
    private byte AVCLevelIndication = 0;
    private byte lengthSizeMinusOneWithReserved = 0;
    private byte numOfSequenceParameterSetsWithReserved = 0;
    private List<byte[]> sequenceParameterSets = null;
    private List<byte[]> pictureParameterSets = null;

    /*
    private byte chromaFormatWithReserve = 0;
    private byte bitDepthLumaMinus8WithReserve = 0;
    private byte bitDepthChromaMinus8WithReserve = 0;
    private List<byte[]> sequenceParameterSetExt = null;
    */

    public AVCConfigurationRecord() {
    }

    public AVCConfigurationRecord(final MediaFormat mediaFormat) {
        // SPS => 0x00,0x00,0x00,0x01,0x67,0x42,0x00,0x29,0x8d,0x8d,0x40,0xa0,0xfd,0x00,0xf0,0x88,0x45,0x38
        ByteBuffer spsBuffer = mediaFormat.getByteBuffer("csd-0");
        // PPS => 0x00,0x00,0x00,0x01,0x68,0xca,0x43,0xc8
        ByteBuffer ppsBuffer = mediaFormat.getByteBuffer("csd-1");

        if (spsBuffer == null || ppsBuffer == null) {
            throw new IllegalStateException();
        }

        /**
         * SPS layout
         **
         profile_idc (8)
         constraint_set0_flag (1)
         constraint_set1_flag (1)
         constraint_set2_flag (1)
         constraint_set3_flag (1)
         constraint_set4_flag (1)
         reserved_zero_3bits (3)
         level_idc (8)
         ...
         */
        spsBuffer.position(4);
        byte[] spsBytes = new byte[spsBuffer.remaining()];
        spsBuffer.get(spsBytes);
        setConfigurationVersion((byte) 0x01);
        setAVCProfileIndication(spsBytes[1]);
        setProfileCompatibility(spsBytes[2]);
        setAVCLevelIndication(spsBytes[3]);
        setLengthSizeMinusOneWithReserved((byte) 0xFF);

        // SPS
        setNumOfSequenceParameterSetsWithReserved((byte) 0xE1);
        List<byte[]> spsList = new ArrayList<byte[]>(1);
        spsList.add(spsBytes);
        setSequenceParameterSets(spsList);

        // PPS
        ppsBuffer.position(4);
        byte[] ppsBytes = new byte[ppsBuffer.remaining()];
        ppsBuffer.get(ppsBytes);
        List<byte[]> ppsList = new ArrayList<byte[]>(1);
        ppsList.add(ppsBytes);
        setPictureParameterSets(ppsList);
    }

    public byte getConfigurationVersion() {
        return configurationVersion;
    }

    public AVCConfigurationRecord setConfigurationVersion(final byte configurationVersion) {
        this.configurationVersion = configurationVersion;
        return this;
    }

    public byte getAVCProfileIndication() {
        return AVCProfileIndication;
    }

    public AVCConfigurationRecord setAVCProfileIndication(final byte AVCProfileIndication) {
        this.AVCProfileIndication = AVCProfileIndication;
        return this;
    }

    public byte getProfileCompatibility() {
        return profileCompatibility;
    }

    public AVCConfigurationRecord setProfileCompatibility(final byte profileCompatibility) {
        this.profileCompatibility = profileCompatibility;
        return this;
    }

    public byte getAVCLevelIndication() {
        return AVCLevelIndication;
    }

    public AVCConfigurationRecord setAVCLevelIndication(final byte AVCLevelIndication) {
        this.AVCLevelIndication = AVCLevelIndication;
        return this;
    }

    public byte getLengthSizeMinusOneWithReserved() {
        return lengthSizeMinusOneWithReserved;
    }

    public AVCConfigurationRecord setLengthSizeMinusOneWithReserved(final byte lengthSizeMinusOneWithReserved) {
        this.lengthSizeMinusOneWithReserved = lengthSizeMinusOneWithReserved;
        return this;
    }

    public byte getNumOfSequenceParameterSetsWithReserved() {
        return numOfSequenceParameterSetsWithReserved;
    }

    public AVCConfigurationRecord setNumOfSequenceParameterSetsWithReserved(final byte numOfSequenceParameterSetsWithReserved) {
        this.numOfSequenceParameterSetsWithReserved = numOfSequenceParameterSetsWithReserved;
        return this;
    }

    public List<byte[]> getSequenceParameterSets() {
        return sequenceParameterSets;
    }

    public AVCConfigurationRecord setSequenceParameterSets(final List<byte[]> sequenceParameterSets) {
        this.sequenceParameterSets = sequenceParameterSets;
        return this;
    }

    public List<byte[]> getPictureParameterSets() {
        return pictureParameterSets;
    }

    public AVCConfigurationRecord setPictureParameterSets(final List<byte[]> pictureParameterSets) {
        this.pictureParameterSets = pictureParameterSets;
        return this;
    }

    public byte getNALULength() {
        return (byte)((getLengthSizeMinusOneWithReserved() >> 6) + 1);
    }

    public ByteBuffer toByteBuffer() {
        int capacity = 5;
        final List<byte[]> sequenceParameterSets = getSequenceParameterSets();
        final List<byte[]> pictureParameterSets = getPictureParameterSets();

        for (byte[] sps : sequenceParameterSets) {
            capacity += 3 + sps.length;
        }
        for (byte[] pps : pictureParameterSets) {
            capacity += 3 + pps.length;
        }

        ByteBuffer buffer = ByteBuffer.allocate(capacity);
        buffer.put(getConfigurationVersion());
        buffer.put(getAVCProfileIndication());
        buffer.put(getProfileCompatibility());
        buffer.put(getAVCLevelIndication());
        buffer.put(getLengthSizeMinusOneWithReserved());

        // SPS
        buffer.put(getNumOfSequenceParameterSetsWithReserved());
        for (byte[] sps : sequenceParameterSets) {
            buffer.putShort((short) sps.length);
            buffer.put(sps);
        }

        // PPS
        buffer.put((byte) pictureParameterSets.size());
        for (byte[] pps : pictureParameterSets) {
            buffer.putShort((short) pps.length);
            buffer.put(pps);
        }

        buffer.flip();

        return buffer;
    }

    public String toString() {
        return ToStringBuilder.reflectionToString(this);
    }
}

File: haishinkit/src/main/java/com/haishinkit/iso/AvcConfigurationRecord.kt
package com.haishinkit.iso

import android.media.MediaCodecInfo
import android.media.MediaFormat
import android.util.Log
import com.haishinkit.codec.CodecOption
import com.haishinkit.codec.VideoCodec
import com.haishinkit.util.toPositiveInt
import org.apache.commons.lang3.builder.ToStringBuilder
import java.nio.ByteBuffer
import java.util.ArrayList

data class AvcConfigurationRecord(
    val configurationVersion: Byte = 0x01,
    val avcProfileIndication: Byte = 0,
    val profileCompatibility: Byte = 0,
    val avcLevelIndication: Byte = 0,
    val lengthSizeMinusOneWithReserved: Byte = 0,
    val numOfSequenceParameterSetsWithReserved: Byte = 0,
    val sequenceParameterSets: List<ByteArray>? = null,
    val pictureParameterSets: List<ByteArray>? = null
) {
    val naluLength: Byte
        get() = ((lengthSizeMinusOneWithReserved.toInt() shr 6) + 1).toByte()

    fun encode(buffer: ByteBuffer): AvcConfigurationRecord {
        buffer.put(configurationVersion)
        buffer.put(avcProfileIndication)
        buffer.put(profileCompatibility)
        buffer.put(avcLevelIndication)
        buffer.put(lengthSizeMinusOneWithReserved)
        // SPS
        buffer.put(numOfSequenceParameterSetsWithReserved)
        sequenceParameterSets?.let {
            for (sps in it) {
                buffer.putShort(sps.size.toShort())
                buffer.put(sps)
            }
        }
        // PPS
        pictureParameterSets?.let {
            buffer.put(it.size.toByte())
            for (pps in it) {
                buffer.putShort(pps.size.toShort())
                buffer.put(pps)
            }
        }
        return this
    }

    fun decode(buffer: ByteBuffer): AvcConfigurationRecord {
        val configurationVersion = buffer.get()
        val avcProfileIndication = buffer.get()
        val profileCompatibility = buffer.get()
        val avcLevelIndication = buffer.get()
        val lengthSizeMinusOneWithReserved = buffer.get()
        val numOfSequenceParameterSetsWithReserved = buffer.get()

        val numOfSequenceParameterSets = numOfSequenceParameterSetsWithReserved.toPositiveInt() and RESERVE_NUM_OF_SEQUENCE_PARAMETER_SETS.inv()
        val sequenceParameterSets = mutableListOf<ByteArray>()
        for (i in 0 until numOfSequenceParameterSets) {
            val bytes = ByteArray(buffer.short.toInt())
            buffer.get(bytes)
            sequenceParameterSets.add(bytes)
        }

        val numPictureParameterSets = buffer.get().toPositiveInt()
        val pictureParameterSets = mutableListOf<ByteArray>()
        for (i in 0 until numPictureParameterSets) {
            val bytes = ByteArray(buffer.short.toInt())
            buffer.get(bytes)
            pictureParameterSets.add(bytes)
        }

        return AvcConfigurationRecord(
            configurationVersion = configurationVersion,
            avcProfileIndication = avcProfileIndication,
            profileCompatibility = profileCompatibility,
            avcLevelIndication = avcLevelIndication,
            lengthSizeMinusOneWithReserved = lengthSizeMinusOneWithReserved,
            numOfSequenceParameterSetsWithReserved = numOfSequenceParameterSetsWithReserved,
            sequenceParameterSets = sequenceParameterSets,
            pictureParameterSets = pictureParameterSets
        )
    }

    internal fun allocate(): ByteBuffer {
        var capacity = 5
        sequenceParameterSets?.let {
            for (sps in it) {
                capacity += 3 + sps.size
            }
        }
        pictureParameterSets?.let {
            for (psp in it) {
                capacity += 3 + psp.size
            }
        }
        return ByteBuffer.allocate(capacity)
    }

    internal fun apply(codec: VideoCodec) {
        when (avcProfileIndication.toInt()) {
            66 -> codec.profile = MediaCodecInfo.CodecProfileLevel.AVCProfileBaseline
            77 -> codec.profile = MediaCodecInfo.CodecProfileLevel.AVCProfileMain
            88 -> codec.profile = MediaCodecInfo.CodecProfileLevel.AVCProfileHigh
        }
        when (avcLevelIndication.toInt()) {
            31 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel31
            32 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel32
            40 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel4
            41 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel41
            42 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel42
            50 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel5
            51 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel51
            52 -> codec.level = MediaCodecInfo.CodecProfileLevel.AVCLevel52
        }
        val options = mutableListOf<CodecOption>()
        val spsBuffer = ByteBuffer.allocate(128)
        sequenceParameterSets?.let {
            for (sps in it) {
                spsBuffer.put(AvcFormatUtils.START_CODE)
                spsBuffer.put(sps)
            }
        }
        spsBuffer.flip()
        options.add(CodecOption(CSD0, spsBuffer))
        val ppsBuffer = ByteBuffer.allocate(128)
        pictureParameterSets?.let {
            for (pps in it) {
                ppsBuffer.put(AvcFormatUtils.START_CODE)
                ppsBuffer.put((pps))
            }
        }
        ppsBuffer.flip()
        options.add(CodecOption(CSD1, ppsBuffer))
        codec.options = options
        Log.i(TAG, "apply value=$this for a videoCodec")
    }

    internal fun toByteBuffer(): ByteBuffer {
        val result = ByteBuffer.allocate(128)
        result.put(0x00).put(0x00).put(0x00).put(0x00)
        sequenceParameterSets?.let {
            for (sps in it) {
                result.put(AvcFormatUtils.START_CODE)
                result.put(sps)
            }
        }
        pictureParameterSets?.let {
            for (pps in it) {
                result.put(AvcFormatUtils.START_CODE)
                result.put(pps)
            }
        }
        return result
    }

    override fun toString(): String {
        return ToStringBuilder.reflectionToString(this)
    }

    @Suppress("unused")
    companion object {
        const val RESERVE_LENGTH_SIZE_MINUS_ONE = 0x3F
        const val RESERVE_NUM_OF_SEQUENCE_PARAMETER_SETS = 0xE0
        const val RESERVE_CHROME_FORMAT = 0xFC
        const val RESERVE_BIT_DEPTH_LUMA_MINUS8 = 0xF8
        const val RESERVE_BIT_DEPTH_CHROME_MINUS8 = 0xF8

        private const val CSD0 = "csd-0"
        private const val CSD1 = "csd-1"

        private var TAG = AvcConfigurationRecord::class.java.simpleName

        internal fun create(mediaFormat: MediaFormat): AvcConfigurationRecord {
            // SPS => 0x00,0x00,0x00,0x01,0x67,0x42,0x00,0x29,0x8d,0x8d,0x40,0xa0,0xfd,0x00,0xf0,0x88,0x45,0x38
            val spsBuffer = mediaFormat.getByteBuffer(CSD0)
            // PPS => 0x00,0x00,0x00,0x01,0x68,0xca,0x43,0xc8
            val ppsBuffer = mediaFormat.getByteBuffer(CSD1)
            if (spsBuffer == null || ppsBuffer == null) {
                throw IllegalStateException()
            }
            return AvcConfigurationRecord(
                configurationVersion = 0x01,
                avcProfileIndication = spsBuffer[5],
                profileCompatibility = spsBuffer[6],
                avcLevelIndication = spsBuffer[7],
                lengthSizeMinusOneWithReserved = 0xFF.toByte(),
                numOfSequenceParameterSetsWithReserved = 0xE1.toByte(),
                sequenceParameterSets = ArrayList<ByteArray>(1).apply {
                    val length = spsBuffer.remaining()
                    add(spsBuffer.array().slice(4 until length).toByteArray())
                },
                pictureParameterSets = ArrayList<ByteArray>(1).apply {
                    val length = ppsBuffer.remaining()
                    add(ppsBuffer.array().slice(4 until length).toByteArray())
                }
            )
        }
    }
}
